docs/serialization.md

```diff
@@ -0,0 +1,102 @@
+# Serialization
+
+
+## Introduction
+
+Histogram serialization has to cover a wide range of formats. As such, we
+describe a form for serialization that covers the metadata structure as
+JSON-like, with a provided JSON-schema. The data (bins and/or variable edges)
+is stored out-of-band in a binary format based on what type of data file you
+are in.  For very small (primarily 1D) histograms, data is allowed inline as
+well.
+
+The following formats are being targeted:
+
+```
+┌────────┐ ┌────────┐ ┌───────┐
+│  ROOT  │ │  HDF5  │ │  ZIP  │
+└────────┘ └────────┘ └───────┘
+```
+
+Other formats can be used as well, assuming they support out-of-band data and
+text attributes or files for the metadata.
+
+## Drawbacks
+
+This structure was based heavily on boost-histogram, but it is intended to be
+general, and can be expanded in the future as needed. As such, the following
+limitations are required:
+
+* Serialization followed by deserialisation may cause axis changes. Axis types
+  may change to an equivalent but less performant axis, growth status will be
+  lost, etc.
+* Metadata must be expressible as JSON. It should also be reasonably sized; some
+  formats like HDF5 may limit the size of attributes to 64K.
+* Floating point variations could be introduced in storages, as the storage
+  format uses a stable but different representation.
+* Axis `name` is only part of the metadata, and is not standardized. This is
+  due to lack of support from boost-histogram.
+
+## Design
+
+The following axes types are supported:
+
+* `"regular"`: A regularly spaced set of even bins. Boost-histogram's "integer"
+  axes maps to this axis as well. Has `upper`, `lower`, `bins`, `underflow`,
```

Comment:
just besides, I am a large fan of the "upper", "lower" naming: I think it's the only combination that works in many different "similar" scenarios and has the same length while expressing what you want ;)